### PR TITLE
Add ServerCtrl web api for shutdown

### PR DESCRIPTION
**Pull request by vadmeste**
Simple web api for shutdown/restart

### CODE REVIEW

File: utils.go [hunk 1]
```diff
@@ -127,6 +135,21 @@ func monitorShutdownSignal() {
...
+				if signal == shutdownRestart {
+					path := os.Args[0]
+					cmdArgs := os.Args[1:]
+					cmd := exec.Command(path, cmdArgs...)
+					cmd.Stdout = os.Stdout
+					cmd.Stderr = os.Stderr
+
+					err := cmd.Start()
+					if err != nil {
+						// FIXME: log to the appropriate engine
```
Comment (harshavardhana) [inline utils.go:149]: you can use `errorIf(err,...)`

File: utils.go [hunk 2]
```diff
@@ -127,6 +135,21 @@ func monitorShutdownSignal() {
 					}
 
 				}
+				// All shutdown callbacks ensure that the server is safely terminated
+				// and any concurrent process could be started again
+				if signal == shutdownRestart {
+					path := os.Args[0]
+					cmdArgs := os.Args[1:]
+					cmd := exec.Command(path, cmdArgs...)
```
Comment (harshavardhana) [inline utils.go:143]: Have you looked at saving the known FD's and restarting? - https://golang.org/pkg/os/#StartProcess

Comment (vadmeste) [inline utils.go:143]: @harshavardhana, why will we need that ?

Comment (harshavardhana) [inline utils.go:143]: > @harshavardhana, why will we need that ?

you need to persist the env for the existing process.  Does this support seamless migration of env's, pwd etc?

File: web-handlers.go [hunk 3]
```diff
@@ -68,6 +68,33 @@ type WebGenericRep struct {
...
+	Status    string
+	UIVersion string `json:"uiVersion"`
+}
+
+// ServerCtrlArgs - server control args
+type ServerCtrlArgs struct {
+	Cmd string `json:"Cmd"`
+}
+
+// ServerInfo - get server info.
```
Comment (harshavardhana) [inline web-handlers.go:82]: Fix the comment.

File: utils.go (hunk 2 above)
Comment (harshavardhana) [inline utils.go:143]: > you need to persist the env for the existing process. Does this support seamless migration of env's, pwd etc?

Also when the browser sends the signal how seamless does JSON-RPC calls work? don't see any explanation in the patch what is the preceding nature. Is this a precursor to 'minio control <command>' ?

**Comment by krishnasrinivas** [comment]
@vadmeste the shutdown handler will have to be registered with net/rpc server (not json-rpc) , i will push the code for healing handler by sunday. you can extend it. sorry for the delay.

File: utils.go (hunk 2 above)
Comment (vadmeste) [inline utils.go:143]: > you need to persist the env for the existing process. Does this support seamless migration of env's, pwd etc?

yes, the same env variables are still accessible after restart, even FDs and already sockets could be transfered too if we will need that later

> Also when the browser sends the signal how seamless does JSON-RPC calls work? don't see any explanation in the patch what is the preceding nature. Is this a precursor to 'minio control <command>' ?

sorry I didn't mention that but as you said, it is an introduction to minio control command

**Pull request closed without merging**